patterns/interpreter.md

```diff
@@ -0,0 +1,182 @@
+# Interpreter
+
+## Description
+
+If a problem occurs very often and requires long and repetitive steps to solve
+it, then the problem instances might be expressed in a simple language and an
+interpreter object could solve it by interpreting the sentences written in this
+simple language.
+
+Basically, for any kind of problems we define:
+
+- a [domain specific language](https://en.wikipedia.org/wiki/Domain-specific_language),
+- a grammar for this language
+- an interpreter that solves the problem instances.
+
+## Motivation
+
+Our goal is to translate simple mathematical expressions into
+postfix expressions (or [Reverse Polish notation](https://en.wikipedia.org/wiki/Reverse_Polish_notation))
+For simplicity, our expressions consist of ten digits `0`, ..., `9` and two
+operations `+`, `-`. For example,
+the expression `2 + 4` is translated into `2 4  +`.
+
+## Context Free Grammars
+
+In order to understand the Interpreter pattern you need to know a little bit
+about formal language theory. If you already know what a grammar is feel free
+to skip this section, otherwise fasten your seat belts.
+
+Basicaly, a [Context Free Grammar (CFG)](https://en.wikipedia.org/wiki/Context-free_grammar)
+defines a set of rules that describe all possible strings over some finite
+set of symbols, formally called alphabet. These strings usually have some well
+defined structure. For example, strings of all balanced parentheses:
+
+```ignore
+(), (()), ()(), ()(()), ..
+```
+
+Formally a CFG consists of terminal symbols, non-terminal symbols, a start
+(non-terminal) symbol, and production rules.
+
+In the previous example we have two terminal symbols `(` and `)`. We define a
+single non-terminal (at the same time the start) symbol `S` and production rules
+as:
+
+```ignore
+S -> SS
+S -> (S)
+S -> ()
+```
+
+So, in order to generate a string with balanced parentheses we repeatedly apply
+production rules. For example, to derive `()()` we apply the following rules:
+
+```ignore
+S -> SS -> ()(S) -> ()()
+```
```

Comment:
I am kinda confused at this part other than the other parts, maybe my CS knowledge isn't that good yet. Maybe I get a little bit but still not quite understand.

Reply:
@pickfire which part is confusing? Is explanation too short/not clear? Or is CFGs is not necessary here?

Reply:
I mean I find it harder to understand, maybe not very clear or because the prerequisite knowledge required is too high, not sure if it is necessary.

Reply:
@pickfire 
> the prerequisite knowledge required is too high

In this section I very briefly describe [Context Free Grammars](https://en.wikipedia.org/wiki/Context-free_grammar). This topic is typically covered in Formal languages or Compilers course, and it is typically given at senior undergrad or graduate CS/CE level. I agree that for a someone who never saw CFG this section isn't enough to convey the whole picture about the subject. We could add a note that it is not essential and just simply a reminder and you can skip it.
 
Next, the reason why I am mentioning, here, CFGs is because the Interpreter pattern is closely related to formal languages, though it is not about writing parsers. Initially I didn't include this section. Later I included it on reviewers' recommendations. 

We could remove this section. What do @simonsan and @MarcoIeni think?

Reply:
Hmm, I know what @pickfire means, but I would actually keep it as is and wait for feedback of a broader mass of people. If there are many questions coming up, we can refactor in the future, answer frequently asked questions etc. 

> We could add a note that it is not essential and just simply a reminder and you can skip it.

Indeed that is a viable approach imho 👍🏽 

Y

Reply:
Removed CFG section.